devito/interfaces.py

```diff
@@ -1,81 +1,257 @@
+    present and given as keyword arguments, since SymPy uses `*args`
+    to (re-)create the dimension arguments of the symbolic function.
+    """
 
     def __new__(cls, *args, **kwargs):
-        return IndexedBase.__new__(cls, args[0], shape=args[1])
+        if cls not in _SymbolCache:
+            name = kwargs.get('name')
+            shape = kwargs.get('shape')
+            if len(args) < 1:
+                args = cls.indices(shape)
+            # Create a new type instance from cls and inject name
+            newcls = type(name, cls.__bases__, dict(cls.__dict__))
+            # Create the new Function object and invoke __init__
+            newobj = Function.__new__(newcls, *args)
+            newobj.__init__(*args, **kwargs)
+            return newobj
+        return Function.__new__(cls, *args)
+
+    def __init__(self):
+        """Initialise from a cached instance by shallow copying __dict__"""
+        original = _SymbolCache[self.__class__]
+        self.__dict__ = original.__dict__.copy()
+
+    @classmethod
+    def _cached(cls):
+        """Test if current class is already in the symbol cache"""
+        return cls in _SymbolCache
+
+    @classmethod
+    def _cache_put(cls, obj):
+        """Store given object instance in symbol cache"""
+        _SymbolCache[cls] = obj
+
+    @classmethod
+    def indices(cls, shape):
+        """Abstract class method to determine the default dimension indices."""
+        raise NotImplementedError('Abstract class `SymbolicData` does not have default indices')
+
+
+class DenseData(SymbolicData):
+    """Data object for spatially varying data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param shape: Shape of the spatial data grid
+    :param dtype: Data type of the buffered data
+
+    Note: :class DenseData: objects are assumed to be constant in time and
+    therefore do not support time derivatives. Use :class TimeData: for
+    time-varying griad data.
+    """
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            # Initialise instance from symbol cache
+            SymbolicData.__init__(self)
+            return
+        else:
+            self.name = kwargs.get('name')
+            self.shape = kwargs.get('shape')
+            self.dtype = kwargs.get('dtype', np.float32)
+            self._data = kwargs.get('_data', None)
+            self.initializer = None
+            # Store new instance in symbol cache
+            self._cache_put(self)
+
+    @classmethod
+    def indices(cls, shape):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        """
+        _indices = [x, y, z]
+        return _indices[:len(shape)]
+
+    @property
+    def indexed(self):
+        """Returns base symbol as sympy.IndexedBase"""
+        return IndexedBase(self.name, shape=self.shape)
```

Comment:
Actually, for now we need this to keep the "old style" expressions from triggering during the auto-conversion in the follow-up branch. This is actually a nice feature too, since it allows users to bypass certain symbolic steps and can thus be used to insert "raw stencils" that do not come from an equation; a kind of lower-level backdoor.